File: app/javascript/src/task/components/task_title.js
import autobind from 'class-autobind';
import PropTypes from 'prop-types';
import React from 'react';

import DeleteButton from 'src/task/components/delete_button';
import TaskEditIcon from 'src/task/components/edit_icon';
import TaskEditTitleForm from 'src/task/containers/edit_title_form';
import timeframeNameMap from 'src/timeframe/name_map';
import {taskShape} from 'src/shapes';

class TaskTitle extends React.Component {
  constructor(props) {
    super(props);
    autobind(this);
  }

  className() {
    let classString = 'col-md-12 task-display';

    if (this.props.task.priority) {
      classString = `${classString} priority-${this.props.task.priority}`;
    }
    if (this.props.task.skip_count >= 15) {
      classString = `${classString} over-skipped`;
    }

    return classString;
  }

  title() {
    return `skip count: ${this.props.task.skip_count}`;
  }

  emblems() {
    if (!this.props.task.repeatSeconds) { return false; }

    return <i className='fa fa-repeat' title='task repeats' />;
  }

  timeframeName() {
    if (!this.props.task.timeframe) { return false; }

    const timeframeName = timeframeNameMap[this.props.task.timeframe];

    return <div className='timeframe'>{timeframeName}</div>;
  }

  render() {
    const {deleteTask, task} = this.props;

    return (
      <div className='row'>
        <div id='task' className={this.className()}>
          <table>
            <tbody>
              <tr>
                <td className='col-xs-1'>
                  {this.timeframeName()}
                  <DeleteButton
                    task={task}
                    deleteTask={deleteTask}
                  />
                </td>
                <td className='col-xs-10 title'>
                  <div className='col-xs-10 col-xs-offset-1'>
                    <TaskEditTitleForm keyPrefix={'focusView'} task={task} />
                  </div>

                  <div className='col-xs-1'>
                    <span className='emblems'>{this.emblems()}</span>
                  </div>
                </td>

                <td className='col-xs-1'>
                  <TaskEditIcon task={task} />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}

TaskTitle.propTypes = {
  deleteTask: PropTypes.func.isRequired,
  task: taskShape.isRequired,
};

export default TaskTitle;

File: app/views/tasks/task_title.rb
# frozen_string_literal: true

module Views
  module Tasks
    class TaskTitle < Views::Base
      register_value_helper :edit_icon
      register_value_helper :focus_task_classes
      register_value_helper :timeframe_label
      register_output_helper :focus_delete_button
      register_output_helper :task_link_to
      register_output_helper :task_repeat_emblem

      def initialize(task:)
        @task = task
        super()
      end

      def view_template
        div(class: "row") do
          div(id: "task", class: focus_task_classes(@task)) do
            table { tbody { tr { cells } } }
          end
        end
      end

      private

      def cells
        timeframe_cell
        title_cell
        td(class: "col-xs-1") { task_link_to(edit_icon, @task) }
      end

      def timeframe_cell
        td(class: "col-xs-1") do
          timeframe_label_div
          focus_delete_button(@task)
        end
      end

      def timeframe_label_div
        return unless @task.timeframe

        div(class: "timeframe") { timeframe_label(@task.timeframe) }
      end

      def title_cell
        td(class: "col-xs-10 title") do
          div(class: "col-xs-10 col-xs-offset-1") do
            render(TitleForm.new(task: @task))
          end
          div(class: "col-xs-1") do
            span(class: "emblems") { task_repeat_emblem(@task) }
          end
        end
      end
    end
  end
end
